src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java

```diff
@@ -207,31 +226,52 @@ public void close(int code) throws IOException {
 
     @Override
     public void close(int code, String reason) throws IOException {
-        byte[] reasonBytes = null;
+        if (outputState == CLOSED) {
+            return;
+        }
+
+        try {
+            stateLock.lock();
 
-        if (code != 0) {
-            // Verify code and reason against RFC 6455.
-            // If code is present, it must equal to 1000 or in range 3000 to 4999
-            if (code != 1000 && (code < 3000 || code > 4999)) {
-                throw new IOException(MSG_INVALID_CLOSE_CODE);
+            if (outputState == CLOSED) {
+                return;
             }
 
-            if (reason != null && reason.length() > 0) {
-                reasonBytes = reason.getBytes(UTF_8);
+            byte[] reasonBytes = null;
+
+            if (code != 0) {
+                // Verify code and reason against RFC 6455.
+                // If code is present, it must equal to 1000 or in range 3000 to 4999
+                if (code != 1000 && (code < 3000 || code > 4999)) {
```

Comment:
This is not correct check as it throws exception for well defined codes (for e.g. many impls send 1002 - protocol error).  In server we do the following, when the close message is created

    public static void validateCloseCode(int statusCode) {
        if (statusCode < 1000 || (statusCode > 1011 && statusCode <= 2999) || statusCode == 1004) {
            throw new IllegalArgumentException("Invalid close code: " + statusCode);
        }
    }

If the close message is created from the wire, we also check for 1005 and 1006 (they are not valid on the wire.) apart from the above.